src/runtime_src/core/pcie/driver/linux/xocl/subdev/flash.c

```diff
@@ -18,13 +19,1109 @@
+	int ret = flash_transaction(flash, cmd, sizeof(cmd), 0, true);
+
+	if (ret || (cmd[1] & 0x1)) // flash device is busy
+		return false;
+
+	return true;
+}
+
+static int flash_get_ID(struct xocl_flash *flash)
+{
+	int i;
+	struct qspi_flash_vendor *vendor = NULL;
+	/*
+	 * Reading flash device vendor ID. Vendor ID is in cmd[1], max vector
+	 * number is in cmd[3] from output.
+	 */
+	u8 cmd[5] = { QSPI_CMD_IDCODE_READ, };
+	int ret = flash_transaction(flash, cmd, sizeof(cmd), 0, true);
+
+	if (ret) {
+		FLASH_ERR(flash, "Can't get flash memory ID, err: %d", ret);
+		return -EINVAL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(vendors); i++) {
+		if (cmd[1] == vendors[i].vendor_id) {
+			vendor = &vendors[i];
+			break;
+		}
+	}
+
+	/* Find out flash vendor and size. */
+	if (vendor == NULL) {
+		FLASH_ERR(flash, "Unknown flash vendor: %d", cmd[1]);
+		return -EINVAL;
+	} else {
+		FLASH_INFO(flash, "Flash vendor: %s", vendor->vendor_name);
+		flash->vendor = vendor;
+	}
+
+	flash->flash_size = vendor->code2sectors(cmd[3]) * (16 * 1024 * 1024);
+	if (flash->flash_size == 0) {
+		FLASH_ERR(flash, "Unknown flash memory size code: %d", cmd[3]);
+		return -EINVAL;
+	} else {
+		FLASH_INFO(flash, "Flash size: %ldMB",
+			flash->flash_size / 1024 / 1024);
+	}
+
+	return 0;
+}
+
+static int flash_enable_write(struct xocl_flash *flash)
+{
+	u8 cmd = QSPI_CMD_WRITE_ENABLE;
+	int ret = flash_transaction(flash, &cmd, 1, 0, false);
+
+	if (ret)
+		FLASH_ERR(flash, "Failed to enable flash write: %d", ret);
+	return ret;
+}
+
+static int flash_set_sector(struct xocl_flash *flash, u8 sector)
+{
+	int ret = 0;
+	u8 cmd[] = { QSPI_CMD_EXTENDED_ADDRESS_REG_WRITE, sector };
+
+	if (sector == flash->qspi_curr_sector)
+		return 0;
+
+	FLASH_DBG(flash, "setting sector to %d", sector);
+
+	ret = flash_enable_write(flash);
+	if (ret)
+		return ret;
+
+	ret = flash_transaction(flash, cmd, sizeof(cmd), 0, false);
+	if (ret) {
+		FLASH_ERR(flash, "Failed to set sector %d: %d", sector, ret);
+		return ret;
+	}
+
+	flash->qspi_curr_sector = sector;
+	return ret;
+}
+
+/* For 24 bit addressing. */
+static inline void flash_offset2faddr(loff_t addr,
+	struct qspi_flash_addr *faddr)
+{
+	faddr->slave = (u8)(addr >> 56);
+	faddr->sector = (u8)(addr >> 24);
+	faddr->addr_lo = (u8)(addr);
+	faddr->addr_mid = (u8)(addr >> 8);
+	faddr->addr_hi = (u8)(addr >> 16);
+}
+
+/* IO cmd starts with op code followed by address. */
+static inline int
+flash_setup_io_cmd_header(struct xocl_flash *flash,
+	u8 op, struct qspi_flash_addr *faddr, size_t *header_len)
+{
+	int ret = 0;
+
+	/* Set sector (the high byte of a 32-bit address), if needed. */
+	ret = flash_set_sector(flash, faddr->sector);
+	if (ret == 0) {
+		/* The rest of address bytes are in cmd. */
+		flash->io_buf[0] = op;
+		flash->io_buf[1] = faddr->addr_hi;
+		flash->io_buf[2] = faddr->addr_mid;
+		flash->io_buf[3] = faddr->addr_lo;
+		*header_len = 4;
+	}
+	return ret;
+}
+
+static bool flash_wait_till_ready(struct xocl_flash *flash)
```

Comment:
Grammatically the method name is a bit weird, till is not correct word, should be until.

Reply:
sure. will change.